Setup py3 test job

mvdbeek opened the pull request:
@jmchilton if the Dockerfile looks OK to you, could you build a new image ?

xref. #1715

mvdbeek committed:
Setup pytest python3 API testing

mvdbeek committed:
Bump docker image version

jmchilton commented on test/docker/base/Dockerfile:
```diff
@@ -79,11 +81,15 @@ RUN mkdir /etc/galaxy && cd /tmp/ansible && mkdir roles && \
+    export GALAXY_VIRTUAL_ENV=$VENV && \
+    . $GALAXY_VIRTUAL_ENV/bin/activate && \
     pip install psycopg2 && \
     pip install mysql && \
```
The loop runs over the rest of this twice but only pre-populates the databases once right? This is fine I'm just double checking - it'd be a little cleaner to end the loop sooner if that is the case but we can merge this as is and do that in a subsequent revision - I know it takes a while to revise these things.

mvdbeek replied:
Good catch, it does call the script twice, I'll change that

jmchilton commented:
Yeah - I can build and publish a new image.

mvdbeek committed:
Avoid prepopulating database twice

mvdbeek committed:
Temporarily run tests with image published in my channel

mvdbeek committed:
Add py3 build script for jenkins

mvdbeek committed:
Explicitly create a python 3 virtualenv

mvdbeek committed:
Bump docker image version

mvdbeek commented:
I'm a bit stuck on this now, the docker image doesn't build because `bdbag` pins `bagit==1.6.4`, and `bagit==1.6.4` can't be installed on python 3.5 -- which has been fixed in bagit 1.7.0:
```
(galaxy_venv3) root@e5e36d3697a2:/# pip install bagit
Requirement already satisfied: bagit in /galaxy_venv3/lib/python3.6/site-packages (1.7.0)
(galaxy_venv3) root@e5e36d3697a2:/# pip install bdbag
Collecting bdbag
  Downloading https://files.pythonhosted.org/packages/53/c1/c28ec526b699813287f94cd4b9971918881505ecb1e7760a3a0fa8d9bf9c/bdbag-1.4.1-py2.py3-none-any.whl (63kB)
    100% |################################| 71kB 838kB/s
Collecting certifi (from bdbag)
  Using cached https://files.pythonhosted.org/packages/7c/e6/92ad559b7192d846975fc916b65f667c7b8c3a32bea7372340bfe9a15fa5/certifi-2018.4.16-py2.py3-none-any.whl
Collecting bagit==1.6.4 (from bdbag)
  Using cached https://files.pythonhosted.org/packages/64/91/4363e2d56a81431ef7b86cd8f0a530a5729032f207eb3b756eb29b4ed282/bagit-1.6.4.tar.gz
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-k6o0335t/bagit/setup.py", line 19, in <module>
        long_description = readme.read()
      File "/galaxy_venv3/lib/python3.6/encodings/ascii.py", line 26, in decode
        return codecs.ascii_decode(input, self.errors)[0]
    UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 1111: ordinal not in range(128)
```

mvdbeek commented:
PR to fix this is [here](https://github.com/fair-research/bdbag/pull/28)

jmchilton commented:
If we have a strict pin on a newer version - does everything fail at that point - I suppose it would? We can make bdbag an optional dependency until it is well behaved with respect to its dependencies in order to make progress. We can just ``@skip`` the relevant tests under Python 3.

mvdbeek commented:
Should be straightforward, we can download the package and do `python setup.py install` until we have a new fixed version. Or is there a way to override the pin in bdbag's setup.py ?

mvdbeek committed:
Fix up Dockerfile for bdbag installation

mvdbeek commented:
On the brightside the pytest setup works (in this instance we tested python 2): https://jenkins.galaxyproject.org/job/docker-api-py3/6/testReport/.

mvdbeek committed:
Use 18.09.0 tag

mvdbeek committed:
Do use python 3

mvdbeek committed:
Decode bytestring returned by source_string

mvdbeek committed:
Decode body when using json.loads in batch middleware

mvdbeek committed:
More resource_string fixes

mvdbeek committed:
Patch LargeBinary to cast strings to bytestring

I don't fully understand the issue, and we might be accumulating
technical debt here. What I'm seeing without this commit is this
when I run for instance the VisualizationsApiTestCase.test_create
API test.

```
Traceback (most recent call last):
  File "/galaxy/lib/galaxy/web/framework/decorators.py", line 283, in decorator
    rval = func(self, trans, *args, **kwargs)
  File "/galaxy/lib/galaxy/webapps/galaxy/api/visualizations.py", line 55, in index
    visualizations = self.get_visualizations_by_user(trans, user)
  File "/galaxy/lib/galaxy/web/base/controller.py", line 664, in get_visualizations_by_user
    return query.all()
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/query.py", line 2737, in all
    return list(self)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/loading.py", line 98, in instances
    util.raise_from_cause(err)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 187, in reraise
    raise value
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/loading.py", line 79, in instances
    rows = [proc(row) for row in fetch]
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/loading.py", line 79, in <listcomp>
    rows = [proc(row) for row in fetch]
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/loading.py", line 493, in _instance
    loaded_instance, populate_existing, populators)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/loading.py", line 604, in _populate_full
    populator(state, dict_, row)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/strategies.py", line 1782, in load_scalar_from_joined_new_row
    dict_[key] = _instance(row)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/loading.py", line 493, in _instance
    loaded_instance, populate_existing, populators)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/orm/loading.py", line 593, in _populate_full
    dict_[key] = getter(row)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/sql/type_api.py", line 1204, in process
    return process_value(impl_processor(value), dialect)
  File "/galaxy_venv3/lib/python3.5/site-packages/sqlalchemy/sql/sqltypes.py", line 902, in process
    value = bytes(value)
TypeError: string argument without an encoding
```

I think LargeBinary isn't meant to store strings the way we do this (we
dump JSON as bytestring), and somehow on python 3 we get unicode strings
back from the database, which blows up here. I suspect this might be
configured on the connection level (see
http://docs.sqlalchemy.org/en/latest/dialects/postgresql.html#unicode),
but not blowing up here seems like a good default.

mvdbeek committed:
Mount in run_tests script at container runtime

This allows changing variables in the script without rebuilding the
container

mvdbeek committed:
Make random_lines1 consistent on py2 and py3 (again)

Turns out integers are used as is in `random.seed()`, but strings are
hashed (in the version 1 protocol), and the hashing isn't stable. So
instead we hack around this by using `ord()` on each character and
concatenating the digits.  Yucky, but probably better than using numpy
and putting this tool on the list of tools that inherit galaxy's
environment (or adding a requirement tag and resolving with conda ...).

mvdbeek commented:
Alright, this is passing now: https://jenkins.galaxyproject.org/job/docker-api-py3/10/
Before merging we'll need a new image and flip back the image to galaxy/testing-base:18.09.0

jmchilton commented:
```
% docker pull 'mvdbeek/testing-base:18.09.0'
18.09.0: Pulling from mvdbeek/testing-base
8ee29e426c26: Pull complete
6e83b260b73b: Pull complete
e26b65fd1143: Pull complete
40dca07f8222: Pull complete
b420ae9e10b3: Pull complete
28dfb9ba5664: Pull complete
930784f0d6fc: Pull complete
1228d52f23a5: Pull complete
8c1f55b45f64: Pull complete
4de36f30db26: Pull complete
c52b1ac944b1: Pull complete
fb32b7d9670c: Pull complete
27ce2b3a0804: Pull complete
df4dde9b6ecd: Pull complete
525367ae10f8: Pull complete
4e60ccbb9a10: Pull complete
10f9fabb48c2: Pull complete
860fd37df1af: Pull complete
d35d59a11e9f: Pull complete
99646ea51eeb: Pull complete
42960f87951d: Pull complete
75378962b5d9: Pull complete
Digest: sha256:e34f9906cdaed7ec28d9627b0e165ba8312f014a355068f14406b6dccb4a1e07
Status: Downloaded newer image for mvdbeek/testing-base:18.09.0
% docker tag 'mvdbeek/testing-base:18.09.0' 'galaxy/testing-base:18.09.0'
% docker push 'galaxy/testing-base:18.09.0'
The push refers to repository [docker.io/galaxy/testing-base]
f4e6c61120a1: Mounted from mvdbeek/testing-base
1c4f959cba00: Mounted from mvdbeek/testing-base
b9ea605c7e6a: Mounted from mvdbeek/testing-base
875ebe897543: Mounted from mvdbeek/testing-base
2c31e2d6b7ec: Mounted from mvdbeek/testing-base
1b7d90784c1a: Mounted from mvdbeek/testing-base
baea7298b045: Mounted from mvdbeek/testing-base
e9d02a2a710a: Mounted from mvdbeek/testing-base
948192940c16: Mounted from mvdbeek/testing-base
f046bf1cba59: Mounted from mvdbeek/testing-base
65b401a9b1c0: Mounted from mvdbeek/testing-base
5da4616b40eb: Mounted from mvdbeek/testing-base
832b541b4b05: Mounted from mvdbeek/testing-base
9517e241e25a: Mounted from mvdbeek/testing-base
e2d2d156739f: Mounted from mvdbeek/testing-base
22ee48af0819: Mounted from mvdbeek/testing-base
5c269795d1e4: Mounted from mvdbeek/testing-base
bcff331e13e3: Mounted from mvdbeek/testing-base
2166dba7c95b: Mounted from mvdbeek/testing-base
5e95929b2798: Mounted from mvdbeek/testing-base
c2af38e6b250: Mounted from mvdbeek/testing-base
0a42ee6ceccb: Mounted from mvdbeek/testing-base
18.09.0: digest: sha256:e34f9906cdaed7ec28d9627b0e165ba8312f014a355068f14406b6dccb4a1e07 size: 4922
```

mvdbeek committed:
Flip testing image back to galaxy namespace

mvdbeek committed:
py3: Fix repository installation

I'm very surprised this actually worked. I suppose we called my system's
`hg` ?!

mvdbeek committed:
Start testing integration tests on python 3

mvdbeek committed:
Fix showing, creating and downloading HTML files

mvdbeek committed:
py3: fix installing and running data managers

mvdbeek committed:
Fix dictify_outputs

The hack of setting `.__len` doesn't work on python 3.
Subclassing OrderedDict and providing list-like access
using integers should have the same effect.

mvdbeek committed:
py3: Fix optional dataset bool value

This fixes cheetah like `#if $input.dataset: ... #end if`.
Which is a little weird, but fastqc uses this pattern.

mvdbeek committed:
py3: Fix uwsgi mule startup

mvdbeek committed:
py3: Fix new direct API-backed uploads

mvdbeek committed:
py3: fix secret decoder script

mvdbeek committed:
py3: Fix mules test cases

We're now just copying the current python path. Also the handler names
have actually changed, I think this hasn't been updated after a late
change (it is now as it should be, based on the documentation).

mvdbeek committed:
Fix GRT export test

mvdbeek committed:
Do test mules on Jenkins, I don't see why this shouldn't work

mvdbeek commented:
Alright, integration tests (including TS installs, all thanks to @nsoranzo who moved the cloning to `hg` calls) should be running as well.

jmchilton commented:
Wow - way to keep up the awesome. This is great! I guess we should wait for the integration tests to pass on Jenkins now that the mule tests are re-enabled - it was working before when running the tests all together but the mule tests would work when they were the only thing running.

mvdbeek commented:
Yeah, I haven't set up the jenkins job since it's down, but it passes inside docker.

mvdbeek commented:
@galaxybot test this

mvdbeek committed:
Fix repository dependency manager

mvdbeek committed:
Deal with docker localhost quirks

mvdbeek committed:
py3: Fix Maf datatype

mvdbeek committed:
Fix stdout/stderr exception handling

mvdbeek commented:
Had to wrestle a little more with `localhost` inside docker, but https://github.com/galaxyproject/galaxy/pull/6565/commits/dd7e4c601a3f31ce9c4fa06afc8234a39f0a5804 solved that. Here's a run of the integration test that passed on py3: 
https://jenkins.galaxyproject.org/job/integration-py3/3/testReport/

jmchilton commented:
Amazing!
